core/inject.js

```diff
@@ -232,71 +232,162 @@ Blockly.createMainWorkspace_ = function(svg, options, blockDragSurface,
-              block.moveBy(0, overflowBottom);
-              movedBlocks = true;
-            }
-            // Bump any block that's off the left back inside.
-            var overflowLeft = MARGIN + edgeLeft -
-                blockXY.x - (options.RTL ? 0 : blockHW.width);
-            if (overflowLeft > 0) {
-              block.moveBy(overflowLeft, 0);
-              movedBlocks = true;
-            }
-            // Bump any block that's off the right back inside.
-            var overflowRight = edgeLeft + metrics.viewWidth - MARGIN -
-                blockXY.x + (options.RTL ? blockHW.width : 0);
-            if (overflowRight < 0) {
-              block.moveBy(overflowRight, 0);
-              movedBlocks = true;
-            }
+
+          switch (e.type) {
+            case Blockly.Events.BLOCK_CREATE:
+            case Blockly.Events.BLOCK_MOVE:
+              var object = mainWorkspace.getBlockById(e.blockId);
+              break;
+            case Blockly.Events.COMMENT_CREATE:
+            case Blockly.Events.COMMENT_MOVE:
+              var object = mainWorkspace.getCommentById(e.commentId);
+              break;
+          }
+          var objectMetrics = getWorkspaceObjectMetrics(object);
+
+          // Bump any object that's above the top back inside.
+          var overflowTop = metrics.viewTop - objectMetrics.top;
+          if (overflowTop > 0) {
+            object.moveBy(0, overflowTop);
           }
+
+          // Bump any object that's below the bottom back inside.
+          var overflowBottom = metrics.viewBottom - objectMetrics.bottom;
+          if (overflowBottom < 0) {
+            object.moveBy(0, overflowBottom);
+          }
+
+          // Bump any object that's off the left back inside.
+          var overflowLeft = metrics.viewLeft - objectMetrics.left;
+          if (overflowLeft > 0) {
+            object.moveBy(overflowLeft, 0);
+          }
+
+          // Bump any object that's off the right back inside.
+          var overflowRight = metrics.viewRight - objectMetrics.right;
+          if (overflowRight < 0) {
+            object.moveBy(overflowRight, 0);
+          }
+
           if (e) {
-            if (!e.group && movedBlocks) {
-              console.log('WARNING: Moved blocks in bounds but there was no event group.'
-                        + ' This may break undo.');
+            if (!e.group) {
+              console.log('WARNING: Moved object in bounds but there was no' +
+                ' event group. This may break undo.');
             }
             Blockly.Events.setGroup(oldGroup);
           }
         }
       }
     };
-    mainWorkspace.addChangeListener(workspaceChanged);
+    mainWorkspace.addChangeListener(bumpObjects);
   }
+
+  // A reminder for developers to add any events that have to do with
+  // objects on/in the workspace to the bumpObjects function. Useful when
+  // testing often happens on a bounded workspace.
+  var bumpObjectsEventChecker = function(e) {
```

Comment:
Do you know of any way to make closure not add this code to the minified files? Or to have it not run in compiled mode?

I added it because people will definitely forget to update this function, especially when most of the testing happens on a moveable workspace. But developers using blockly don't need to see this message.

Reply:
I don't think we need this function.  Adding events isn't a super common thing yet, and this is the sort of information that we should include in the documentation on how to create a new event type instead of having it be a runtime check.

Rather than checking directly against the event types in your change handler, you can make a list in Blockly.Events of event types that should be bumped, and check the event type against that.  That way you don't have to enumerate all of the event types all over the place, and that code will be easier to read.

You can put a comment on that list saying that anyone adding a new event type that will cause bumping should add it to the list.